benchmark/benchmark_operations.py

```diff
@@ -0,0 +1,110 @@
+from pyrsistent import PClass, field
+from zope.interface import Interface, implementer
+
+from twisted.internet.defer import succeed, maybeDeferred
+from twisted.web.client import ResponseFailed
+
+
+class IProbe(Interface):
+    """A probe that performs an operation."""
+
+    def run():
+        """
+        Run the operation.  This should run with as little overhead as
+        possible, in order to ensure benchmark measurements are accurate.
+
+        :return: A Deferred firing with the result of the operation.
+        """
+
+    def cleanup():
+        """
+        Perform any cleanup required after the operation.  This is performed
+        outside the benchmark measurement.
+
+        :return: A Deferred firing when the cleanup is finished.
+        """
+
+
+class IOperation(Interface):
+    """An operation that can be performed."""
+
+    def get_probe():
+        """
+        Get a probe for the operation. To ensure sequential operations
+        perform real work, the operation may return a different
+        probe each time.
+        """
+
+
+def _report_error(failure):
+    failure.trap(ResponseFailed)
+    for reason in failure.value.reasons:
+        reason.printTraceback()
+    return reason
+
+
+@implementer(IProbe)
+class _NoOpRequest(PClass):
+    """
+    A probe that performs no operation.
+    """
+
+    def run(self):
+        return succeed(None)
+
+    def cleanup(self):
+        return succeed(None)
+
+
+@implementer(IOperation)
+class _NoOperation(PClass):
+    """
+    An nop operation.
+    """
+
+    client = field(mandatory=True)
+
+    def get_probe(self):
+        return _NoOpRequest()
+
+
+@implementer(IProbe)
+class _ReadRequest(PClass):
+    """
+    A probe to perform a read request on the control service.
+    """
+
+    client = field(mandatory=True)
+
+    def run(self):
+        d = self.client.list_datasets_state()
+        d.addErrback(_report_error)
+        return d
+
+    def cleanup(self):
+        return succeed(None)
+
+
+@implementer(IOperation)
+class _ReadRequestOperation(PClass):
+    """
+    An operation to perform a read request on the control service.
+    """
+
+    client = field(mandatory=True)
+
+    def get_probe(self):
+        return _ReadRequest(client=self.client)
+
+
+_operations = {
+    'nop': _NoOperation,
+    'read-request': _ReadRequestOperation,
+}
+
+supported_operations = _operations.keys()
+default_operation = 'read-request'
+
+
+def get_operation(client, name):
+    return maybeDeferred(_operations[name], client=client)
```

Comment:
:grey_question: This pattern (registry, supported list, default, get by name) is repeated three times, which is normally my trigger point for refactoring. I think it's simple enough to leave as-is, but if you wanted to make a `Registry` object or something I wouldn't stop you. (If you do though, I'd like a quick look before merging).